File: web/app/utils/location.py
"""
Location normalization for flexible search.

Transforms a location string into a set of tokens with aliases:
- Canadian provinces (on ↔ ontario, bc ↔ british columbia)
- GTA cities (mississauga, brampton → adds gta, toronto)

Used when searching listings (services/listings.py).
"""

import re


def build_location_search(location: str) -> str:
    """Normalizes location for more flexible searching.

    Adds province aliases and simple local expansions.
    """
    if not location:
        return ""

    raw = location.strip().lower()

    province_aliases = {
        "on": "ontario",
        "ontario": "on",
        "bc": "british columbia",
        "british columbia": "bc",
        "ab": "alberta",
        "alberta": "ab",
        "qc": "quebec",
        "quebec": "qc",
        "mb": "manitoba",
        "manitoba": "mb",
        "sk": "saskatchewan",
        "saskatchewan": "sk",
        "ns": "nova scotia",
        "nova scotia": "ns",
        "nb": "new brunswick",
        "new brunswick": "nb",
        "pe": "prince edward island",
        "prince edward island": "pe",
        "nl": "newfoundland labrador",
        "newfoundland": "nl",
        "newfoundland labrador": "nl",
    }

    clean = re.sub(r"[^a-z0-9\s,]", " ", raw)
    parts = [p for p in re.split(r"[\s,]+", clean) if p]

    expanded = set(parts)

    joined = " ".join(parts)
    if joined in province_aliases:
        expanded.add(province_aliases[joined])

    for part in parts:
        if part in province_aliases:
            expanded.add(province_aliases[part])

    if {"mississauga", "toronto", "brampton", "vaughan"} & expanded:
        expanded.add("gta")
        expanded.add("greater toronto area")

    return " ".join(sorted(expanded))

File: web/internal/utils/location.go
package utils

import (
	"sort"
	"strings"
)

var provinceAliases = map[string]string{
	"on": "ontario",
	"ont": "ontario",
	"bc": "british columbia",
	"ab": "alberta",
	"sk": "saskatchewan",
	"mb": "manitoba",
	"qc": "quebec",
	"nb": "new brunswick",
	"ns": "nova scotia",
	"pe": "prince edward island",
	"nl": "newfoundland and labrador",
	"yt": "yukon",
	"nt": "northwest territories",
	"nu": "nunavut",
}

var gtaCities = []string{"mississauga", "toronto", "brampton", "vaughan", "markham", "richmond hill", "oakville", "burlington", "ajax", "pickering", "oshawa", "whitby"}

// BuildLocationSearch normalizes a location string into searchable tokens.
func BuildLocationSearch(location string) string {
	lower := strings.ToLower(strings.TrimSpace(location))
	tokens := make(map[string]struct{})

	for _, part := range strings.FieldsFunc(lower, func(r rune) bool {
		return r == ' ' || r == ',' || r == '-' || r == '/'
	}) {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		tokens[part] = struct{}{}
		if alias, ok := provinceAliases[part]; ok {
			tokens[alias] = struct{}{}
		}
	}

	// Add GTA synonyms
	for city := range tokens {
		for _, gta := range gtaCities {
			if city == gta {
				tokens["gta"] = struct{}{}
				tokens["greater"] = struct{}{}
				tokens["toronto"] = struct{}{}
				tokens["area"] = struct{}{}
				break
			}
		}
	}

	var result []string
	for t := range tokens {
		result = append(result, t)
	}
	sort.Strings(result)
	return strings.Join(result, " ")
}
